### PR TITLE
Stored Bee API

### PR DESCRIPTION
**Pull request by EOT3000**
Adds a way to modify some properties of individual stored entities (bees in beehive), as well as release them individually, without having to release them all.

With requested changes.

### CODE REVIEW

File: patches/server/0307-Stored-Bee-API.patch [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
++
++        return list;
++    }
++
++    public List<Entity> releaseBee(BlockState iblockdata, Level level, BeeData data, BeehiveBlockEntity.BeeReleaseStatus tileentitybeehive_releasestatus, boolean force) {
++        List<Entity> list = Lists.newArrayList();
++
++        BeehiveBlockEntity.releaseBee(level, this.worldPosition, iblockdata, data, list, tileentitybeehive_releasestatus, this.savedFlowerPos, force);
++
++        if (!list.isEmpty()) {
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Is the list ever modified?

File: patches/server/0307-Stored-Bee-API.patch [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+ public class CraftBeehive extends CraftBlockEntityState<BeehiveBlockEntity> implements Beehive {
+ 
++    private final List<org.purpurmc.purpur.entity.StoredEntity<Bee>> storage = new ArrayList<>(); // Purpur
++
+     public CraftBeehive(World world, BeehiveBlockEntity tileEntity) {
+         super(world, tileEntity);
++
++        // Purpur start - load bees to be able to modify them individually
++
++        loadInitialBees();
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: This function call can be inlined

File: patches/server/0307-Stored-Bee-API.patch [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+     }
+ 
+     @Override
+@@ -66,25 +80,75 @@ public class CraftBeehive extends CraftBlockEntityState<BeehiveBlockEntity> impl
+         List<Bee> bees = new ArrayList<>();
+ 
+         if (isPlaced()) {
+-            BeehiveBlockEntity beehive = ((BeehiveBlockEntity) this.getTileEntityFromWorld());
+-            for (Entity bee : beehive.releaseBees(this.getHandle(), BeeReleaseStatus.BEE_RELEASED, true)) {
++            // Purpur - change which releaseBees method is called, and use beehive snapshot
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Use `// Purpur start` and `// Purpur end` for multiline changes

File: patches/server/0307-Stored-Bee-API.patch [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+             }
+         }
+ 
++        storage.clear(); // Purpur
++
+         return bees;
+     }
+ 
++    // Purpur start
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
++        }
++
++        return null;
++    }
++
++    @Override
++    public List<org.purpurmc.purpur.entity.StoredEntity<Bee>> getEntities() {
++        return new ArrayList<>(storage);
++    }
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+     @Override
+     public void addEntity(Bee entity) {
+         Preconditions.checkArgument(entity != null, "Entity must not be null");
+ 
++        int length = getSnapshot().getStored().size();  // Purpur
++
+         getSnapshot().addOccupant(((CraftBee) entity).getHandle(), false);
++
++        // Purpur start - check if new bee was added, and if yes, add to stored bees
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
++        List<BeehiveBlockEntity.BeeData> s = getSnapshot().getStored();
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
++            storage.add(new org.purpurmc.purpur.entity.PurpurStoredBee(s.get(s.size() - 1), this));
++        }
++
++        // Purpur end
+     }
+     // Paper start - Add EntityBlockStorage clearEntities
+     @Override
+     public void clearEntities() {
+         getSnapshot().clearBees();
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
++        // Purpur start - load bees to be able to modify them individually
++
++        loadInitialBees();
++    }
++
++    private void loadInitialBees() {
++        for(BeehiveBlockEntity.BeeData data : getSnapshot().getStored()) {
++            storage.add(new org.purpurmc.purpur.entity.PurpurStoredBee(data, this));
++        }
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+ 
+ public class CraftBeehive extends CraftBlockEntityState<BeehiveBlockEntity> implements Beehive {
+ 
++    private final List<org.purpurmc.purpur.entity.StoredEntity<Bee>> storage = new ArrayList<>(); // Purpur
++
+     public CraftBeehive(World world, BeehiveBlockEntity tileEntity) {
+         super(world, tileEntity);
++
++        // Purpur start - load bees to be able to modify them individually
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/api/0051-Stored-Bee-API.patch [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
++    T releaseEntity(@NotNull org.purpurmc.purpur.entity.StoredEntity<T> entity);
++
++    /**
++     * Gets all the entities currently stored in the block.
++     *
++     * @return List of all entities which are stored in the block
++     */
++    @NotNull
++    List<org.purpurmc.purpur.entity.StoredEntity<T>> getEntities();
++
```
Comment (katekerllenevich) [inline patches/api/0051-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
++    }
++
++    // Purpur end
++
+     @Override
+     public void addEntity(Bee entity) {
+         Preconditions.checkArgument(entity != null, "Entity must not be null");
+ 
++        int length = getSnapshot().getStored().size();  // Purpur
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

File: patches/server/0307-Stored-Bee-API.patch [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+-            for (Entity bee : beehive.releaseBees(this.getHandle(), BeeReleaseStatus.BEE_RELEASED, true)) {
++            // Purpur - change which releaseBees method is called, and use beehive snapshot
++            BeehiveBlockEntity beehive = ((BeehiveBlockEntity) this.getSnapshot());
++            for (Entity bee : beehive.releaseBees(this.getHandle(), getWorldHandle().getMinecraftWorld(), BeeReleaseStatus.BEE_RELEASED, true)) {
+                 bees.add((Bee) bee.getBukkitEntity());
+             }
+         }
+ 
++        storage.clear(); // Purpur
++
```
Comment (katekerllenevich) [inline patches/server/0307-Stored-Bee-API.patch]: Small nitpick, but remove extra whitespace

**Review by katekerllenevich — CHANGES_REQUESTED** [review]
The whitespace stuff is largely just to make the diff smaller and more maintainable for us, nothing against your code style.

File: patches/server/0307-Stored-Bee-API.patch (hunk 1 above)
Comment (EOT3000) [reply patches/server/0307-Stored-Bee-API.patch]: List is passed to BeehiveBlockEntity.releaseBee, and the vanilla method will modify it.

**Comment by EOT3000** [comment]
Done.

**Review by granny — APPROVED** [review]
LGTM. For future reference make sure to create a new branch specifically for the PR you're creating instead of committing directly to the `ver/...` branch.

**Pull request merged**